### PR TITLE
bambootracker: init at 0.4.3

### PR DESCRIPTION
**Pull request by OPNA2608**
###### Motivation for this change
Been in my local checkout for awhile, decided to push since it was mentioned in https://github.com/NixOS/nixpkgs/issues/81815.

Ping @fgaz because you were interested in packaging it. Give it a look, let me know how it looks to you.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after) (new package -> 326780408)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by OPNA2608** [comment]
Seems to eval fine on x86 & ARM Linux, maybe a trusted user could kick off a Darwin build for completeness' sake? :slightly_smiling_face:

File: pkgs/applications/audio/bambootracker/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  version = "0.3.5";
+
+  src = fetchFromGitHub {
+    owner = "rerrahkr";
+    repo = "BambooTracker";
+    rev = "v${version}";
+    sha256 = "06ij0k5sfns1zakl6qbxr5j91dj1kfj3s5jc3mpqzxc3zpc3avkv";
+  };
+
+  preConfigure = "cd BambooTracker";
```
Comment (fgaz) [inline pkgs/applications/audio/bambootracker/default.nix]: Why not `sourceRoot`?

Comment (OPNA2608) [reply pkgs/applications/audio/bambootracker/default.nix]: Good point, fixed.

File: pkgs/applications/audio/bambootracker/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  ++ opt jackSupport  "DEFINES+=__UNIX_JACK__";
+
+  postInstall = ''
+    install -d $out/share/{man/man1,applications,BambooTracker,icons/hicolor/{16x16,256x256}/apps}
+    install -t $out/share/man/man1/      ../BambooTracker*.1
+    install -t $out/share/applications/  ../BambooTracker.desktop
+    cp -at     $out/share/BambooTracker/ ../demos
+
+    for size in 16x16 256x256; do
+      install res/icon/icon_$size.png $out/share/icons/hicolor/$size/apps/BambooTracker.png
```
Comment (fgaz) [inline pkgs/applications/audio/bambootracker/default.nix]: `install -m644` on non-executables.

You may also want to use `-D` and skip the `-d` step, so those lines don't need to be kept in sync

Comment (OPNA2608) [reply pkgs/applications/audio/bambootracker/default.nix]: Gonna change it to `install -Dm644` for the icons & desktop file.

I saw `installManPage` being used in a recent PR, should I use that instead for the manpages?

For the demos, I don't know how to `install` a whole directory with subdirectories. Should I just `chmod 644 -R $out/share/BambooTracker/demos` after `cp`ing them over?

Comment (fgaz) [reply pkgs/applications/audio/bambootracker/default.nix]: > installManPage

I never had to install a manpage yet, but I guess if something like this exists, using it is a goo idea.

> a whole directory
Just `cp -r` is fine afaik. No need to chmod unless the dev set them wrong in the first place I think.

Comment (OPNA2608) [reply pkgs/applications/audio/bambootracker/default.nix]: Fixed.

File: pkgs/applications/audio/bambootracker/default.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  qmakeFlags = []
+  ++ opt alsaSupport  "DEFINES+=__LINUX_ALSA__"
+  ++ opt pulseSupport "DEFINES+=__LINUX_PULSE__"
+  ++ opt jackSupport  "DEFINES+=__UNIX_JACK__";
+
+  postInstall = ''
+    install -Dm644 ../BambooTracker.desktop $out/share/applications/BambooTracker.desktop
+    installManPage ../BambooTracker*.1
+    cp -r ../demos $out/share/BambooTracker/
```
Comment (fgaz) [inline pkgs/applications/audio/bambootracker/default.nix]: imo keeping the `demos` subdirectory (`cp -r ../demos $out/share/BambooTracker/demos`) would be a bit nicer than having its contents copied directly in `share/BamboTracker`, but this is fine too

**Review by fgaz — APPROVED** [review]

File: pkgs/applications/audio/bambootracker/default.nix (hunk 3 above)
Comment (OPNA2608) [reply pkgs/applications/audio/bambootracker/default.nix]: Whoops, that was my intention. Fixed.

**Comment by OPNA2608** [comment]
Updated to new v0.4.0 release.

**Review by fgaz — APPROVED** [review]

**Comment by OPNA2608** [comment]
Noticed that I missed the i18n & skin files, added those. Does copying all the licensing files to `share/BambooTracker` make sense or is that abit too excessive?

**Comment by fgaz** [comment]
I'd like to know that too. In my packages I didn't bother to include them if the makefile/whatever didn't do it by itself, but maybe I should have.

**Comment by OPNA2608** [comment]
Updated to new v0.4.1 release.

We added a workaround for the missing installation of i18n files on our side (it's caused by nixpkgs' qtbase version, QTBUG-77398), so I removed that part of the postInstall phase.

**Comment by OPNA2608** [comment]
Bumped again to v0.4.2.

**Comment by OPNA2608** [comment]
Bumped to v0.4.3.

**Comment by fgaz** [comment]
Let's see if this attracts some committer :)

**Comment by OPNA2608** [comment]
~~Thanks for reminding me that this PR is still open.~~

I added a patch to fix the Darwin build, it can be dropped with the next release. Additionally, I ran `nixpkgs-fmt` on the whole file.

File: pkgs/applications/audio/bambootracker/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    cp -r ../{demos,licenses,skins,LICENSE} $out/share/BambooTracker/
+
+    for size in 16x16 256x256; do
+      install -Dm644 res/icon/icon_$size.png $out/share/icons/hicolor/$size/apps/BambooTracker.png
+    done
+  '';
+
+  meta = with stdenv.lib; {
+    description = "A tracker for YM2608 (OPNA) which was used in NEC PC-8801/9801 series computers.";
```
Comment (danieldk) [inline pkgs/applications/audio/bambootracker/default.nix:56]:
```suggestion
    description = "A tracker for YM2608 (OPNA) which was used in NEC PC-8801/9801 series computers";
```

Description should not end in a dot (following the contribution guidelines).

**Review by fgaz — APPROVED** [review]

File: pkgs/applications/audio/bambootracker/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    homepage = "https://github.com/rerrahkr/BambooTracker";
+    license = licenses.gpl2;
```
Comment (danieldk) [inline pkgs/applications/audio/bambootracker/default.nix:58]: Should be

```suggestion
    license = licenses.gpl2Only;
```

if this is GPL2 only or `gpl2Later` if it's GPL2 or any later version of the GPL.

File: pkgs/applications/audio/bambootracker/default.nix [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  # Fix macOS build until new release
+  patches = [(fetchpatch {
+    url = "https://github.com/rerrahkr/BambooTracker/commit/45346ed99559d44c2e32a5c6138a0835b212e875.patch";
+    sha256 = "1xkiqira1kpcqkacycy0y7qm1brhf89amliv42byijl4palmykh2";
+  })];
+
+  postPatch = "cd BambooTracker";
+
+  # sourceRoot = "source/BambooTracker";
```
Comment (danieldk) [inline pkgs/applications/audio/bambootracker/default.nix:31]: Any reason `sourceRoot` is not used?

**Review by danieldk — CHANGES_REQUESTED** [review]
Jikes, I am sorry that merging this took so long! I added some small comments so that the derivation follows the guidelines. It would also be nice if you could squash the two commits into one.

Feel free to ping me when you have incorporated these changes, so that we can get this merged!

File: pkgs/applications/audio/bambootracker/default.nix (hunk 6 above)
Comment (OPNA2608) [reply pkgs/applications/audio/bambootracker/default.nix]: Makes patch from GitHub commit fail.

```
patching sources
applying patch /nix/store/05w1rpcw9j4y1wik5bqblhc4vnzhmdkv-45346ed99559d44c2e32a5c6138a0835b212e875.patch
can't find file to patch at input line 3
```

**Comment by OPNA2608** [comment]
@danieldk 

- Rebased so the new license value can be used (this predated that change by months)
- Squashed commits
- Fixed the extra dot

Can't use `sourceRoot` due to the patch.

**Comment by OPNA2608** [comment]
Am I doing something wrong btw? GitHub CI is flooding me with "Run failed" emails.

**Comment by danieldk** [comment]
Thanks for making the changes! Not sure where the GitHub CI errors are coming from. But ofborg is happy.

**Comment by OPNA2608** [comment]
Thanks for merging! :+1:

**Pull request merged**